### ISSUE TITLE
[DOOR Bug] `numpy.typing` Support Bugs

### ISSUE THREAD

**Issue by wesselb** [beartype/beartype#212]
Hey @leycec! I'm opening an issue to report an issue with `TypeHint` originally reported by @francesco-ballarin in [this issue](https://github.com/wesselb/plum/issues/74).

Python 3.11, `TypeHint` appears to complain when it receives a type hint from `numpy.typing`:

```python
>>> TypeHint(npt.NDArray[int])
BeartypeDoorNonpepException: Type hint numpy.ndarray[typing.Any, numpy.dtype[int]] invalid (i.e., either PEP-noncompliant or PEP-compliant but currently unsupported by "beartype.door.TypeHint").
```

On Python 3.9, `TypeHint` happily accepts a `numpy.typing` type hint, but the comparisons aren't quite right:

```python
>>> TypeHint(npt.NDArray[int])  # Yay!
TypeHint(numpy.ndarray[typing.Any, numpy.dtype[int]]) 

>>> TypeHint(npt.NDArray[int]) == TypeHint(npt.NDArray[float])  # :(
True
```

I'm aware that non-`typing` type hints likely won't work out of the box and might need additional support by @beartype, but I thought that it would be helpful to report these issues nevertheless, since `numpy.typing` is widely used.

**Comment by francesco-ballarin** [comment]
```
from numpy import floating, integer
from numpy.typing import NDArray
from beartype.door._cls.doorsuper import TypeHint

TypeHint(NDArray[integer]) == TypeHint(NDArray[floating])
```
gives 
```
Traceback (most recent call last):
  File "/tmp/test.py", line 5, in <module>
    TypeHint(NDArray[integer]) == TypeHint(NDArray[floating])
    ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/beartype/door/_doormeta.py", line 148, in __call__
    _HINT_KEY_TO_WRAPPER.cache_or_get_cached_func_return_passed_arg(
  File "/usr/local/lib/python3.11/dist-packages/beartype/_util/cache/map/utilmapbig.py", line 231, in cache_or_get_cached_func_return_passed_arg
    value = value_factory(arg)
            ^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/beartype/door/_doormeta.py", line 220, in _make_wrapper
    wrapper_subclass = get_typehint_subclass(hint)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/beartype/door/_doordata.py", line 108, in get_typehint_subclass
    raise BeartypeDoorNonpepException(
beartype.roar.BeartypeDoorNonpepException: Type hint numpy.ndarray[typing.Any, numpy.dtype[numpy.integer]] invalid (i.e., either PEP-noncompliant or PEP-compliant but currently unsupported by "beartype.door.TypeHint").
```
with python 3.11 on debian

**Comment by wesselb** [comment]
@tbsexton, ooooh, you're completely right! I definitely did not read the docs as carefully as I should have. Let me adjust the title and the description of the PR.

**Comment by leycec** [comment]
**Gah, gah, gah!** The truth has finally been exposed for all to see. Yes, it's sadly all too true.

Although @beartype has long supported `numpy.typing.NDArray[...]`-style type hints *when type-checking,* <sup>curse that fine print</sup> the `beartype.door.TypeHint` API that scratches Plum's back currently lacks similar support. Let's earmark this for @beartype's upcoming `0.13.0` release, because everyone clearly wants this. So do we!

This topic actually invites a fascinating litany of intriguing questions that will burst everyone's brains. Like, what *is* the correct approach to collating (comparing) NumPy type hints anyway? Nobody knows. I'm pretty sure that:

* NumPy devs are blissfully unaware that a rebellious ~~youth~~ movement of *avant-garde* outsider typing enthusiasts are secretly trying to collate their type hints. (I think I'm talking about us.)
* Everyone except our secret society is equally unaware that type hints even *can* be meaningfully collated.

@beartype kinda *did* accidentally invent this whole glorious conceit six months ago. It was all @tlambert03's doing. But... what do we do now? If anyone has ideas about how to meaningfully collate NumPy dtypes in particular, I will gladly listen and do everything you say.

`int` and `float` aren't actual dtypes, for example; they're just losslessly convertible into dtypes. With that in mind, a few governing principles for the tireless work we do:

* **We only collate actual dtypes.** If somebody subscripts `numpy.typing.NDArray` with anything other than an actual dtype, @beartype silently coerces that into a dtype *before* collation.
* Once an actual dtype is obtained, we:
  * Implement `==`-driven equality comparison in the standard way. That is to say, we just defer to NumPy and hope that their `dtype` class overrides the `__eq__()` dunder method to implement meaningful equality comparison between dtypes. Pretty sure it does – but not certain. **EDIT:** indeed, [NumPy supports this out-of-the-box](https://stackoverflow.com/a/10887287/2809027). We now give praise to the Great NumPy Elder Gods.
  * Implement `<=`-driven subhint comparison by simply deferring to the [existing `numpy.issubdtype()` function](https://numpy.org/doc/stable/reference/generated/numpy.issubdtype.html). That appears to do everything we want to do, based on [this well-researched StackOverflow answer](https://stackoverflow.com/a/10887363/2809027).

If anyone would like to backstab this before I get there, I'd be happy to shepherd you through the ~~gruelling~~ wonderfully easy process of adding a new concrete `beartype.door.NumpyNDArrayTypeHint` subclass. Implement this collation for us and we will shower you with Canadian gold. I am, of course, referring to locally sourced maple syrup – the entire basis of the Canadian economy. :honey_pot:

**Comment by leycec** [comment]
@skumargupta-jumptrading at closed-out issue #42 wants this, too! <sup>...*heh. 42.*</sup> Sadly, it's already summer and I've done nothing about this. This is the worst case.

My current priority for this month is [finalizing our `beartype.claw` API for import hooks](https://github.com/beartype/beartype/issues/43). As soon as that's shipped, I'd be delighted to have a go at this next. Alternately, would anyone else like me to gently mentor them through a pull request (PR)? I assume not, but... a GitHubber's gotta PR-beg. This is the best case.

And thanks to everyone for sticking with @beartype these many years. I promise you that @beartype will someday satisfy your *every* desire. The end goal is in sight!

**Comment by PhilipVinc** [comment]
@leycec is this supported somehow? 
Unfortunately parametric types and generics (dict[a,b]) are still not supported by plum because of this limitation..

**Comment by leycec** [comment]
**Gah!** So sorry about this longstanding deficiency, @wesselb, @PhilipVinc, and Plum crew. Indeed, your SpideySense™ is correct. This is still undone.

Thankfully, I'm a *super*-noice dude for various definitions of "noice." I'd be delighted to walk someone through a PR improving the `TypeHint` API to support all of this fun stuff Plum really needs – including:

* NumPy tensors via `numpy.typing` and/or @patrick-kidger's superb `jaxtyping`.
* JAX tensors via `jax.typing` and/or @patrick-kidger's superb `jaxtyping`.
* PyTorch tensors via... @patrick-kidger's superb `jaxtyping`. I think that's the only way, currently? Kinda sad. Honestly. What you doin', PyTorch? :sob: 
* Parametric types.
* Generics subscripted by two or more types.
* Whatevah else you got for me. :smile: 

You are now thinking: "But I hate the @beartype codebase. It is autistic."

Fear not! Although the @beartype codebase is indeed autistic, the `beartype.door` subpackage is actually quite sensible. It's mostly object-oriented, mostly recursive, and mostly intuitive. Crucially, this means that **it's mostly possible for real humans to improve `beartype.door` and submit those improvements as PRs.**

You are now thinking: "Can't you just do all this stuff for us, @leycec?"

*...heh.* You're not wrong. I could. But @leycec time is limited and mostly just [monotropically](https://en.wikipedia.org/wiki/Monotropism) monofocused on **deep `O(1)` type-checking.** I *do* hope to finish deep `O(1)` type-checking by the end of the year, though. If nobody else hops onto this feature request by then, I'll happily throw myself wholeheartedly onto this flaming pile of fun at that point.

## tl;dr

Patience *or* please PR. These are the unpleasant solutions before us all. Please choose one.
